composer/trainer/scale_schedule.py

```diff
@@ -1,15 +1,50 @@
 # Copyright 2021 MosaicML. All Rights Reserved.
 
+import functools
 from collections import Counter
-from typing import Optional
+from typing import Callable, Optional, Union, get_type_hints
 
 from torch.optim.lr_scheduler import CosineAnnealingLR, CosineAnnealingWarmRestarts, ExponentialLR, MultiStepLR, StepLR
 
+from composer.core.scheduler import ComposerSchedulerFn
 from composer.core.types import Scheduler
-from composer.optim.scheduler import ConstantLR
 
 
-def scale_scheduler(scheduler: Scheduler, ssr: float, orig_max_epochs: Optional[int] = None):
+def _scale_pytorch_scheduler(scheduler: Scheduler, ssr: float, orig_max_epochs: Optional[int] = None):
+    if isinstance(scheduler, StepLR):
+        scheduler.step_size = int(scheduler.step_size * ssr)  # type: ignore  -- unknown attribute
+    elif isinstance(scheduler, MultiStepLR):
+        milestones = scheduler.milestones  # type: ignore  -- unknown attribute
+        milestones = Counter([int(ms * ssr) for ms in milestones])
+        scheduler.milestones = milestones  # type: ignore  -- unknown attribute
+    elif isinstance(scheduler, CosineAnnealingLR):
+        assert orig_max_epochs is not None, "To scale Cosine decay, max_epochs must be provided."
+
+        if hasattr(scheduler, 'interval') and scheduler.interval == "step":  # type: ignore  -- unknown attribute
+            orig_max_epochs *= scheduler.steps_per_epoch  # type: ignore  -- unknown attribute
+
+        warmup = orig_max_epochs - scheduler.T_max  # type: ignore  -- unknown attribute
+        scheduler.T_max = int(orig_max_epochs * ssr - warmup)  # type: ignore  -- unknown attribute
+    elif isinstance(scheduler, CosineAnnealingWarmRestarts):
+        # TODO: account for warmups
+        scheduler.T_0 = int(scheduler.T_0 * ssr)  # type: ignore  -- unknown attribute
+    elif isinstance(scheduler, ExponentialLR):
+        factor = 1 / ssr
+        scheduler.gamma = scheduler.gamma**factor  # type: ignore  -- unknown attribute
+    else:
+        raise ValueError(f'Scale schedule being applied to unrecognized Scheduler {scheduler}. '
+                         'Please implement your scheduler as a function instead.')
+
+    return scheduler
+
+
+def _scale_composer_scheduler(scheduler: ComposerSchedulerFn, ssr: float):
+    return functools.partial(scheduler, ssr=ssr)  # type: ignore
```

Comment:
This is another thing I'm not a huge fan of. It's absolutely necessary for `ssr` to be a scheduler fn param, but I wasn't sure whether I wanted that to be reflected within the `ComposerSchedulerFn` type. I went with leaving the `ssr` param out, which is why the `# type: ignore` is necessary here.

Reply:
I don't think having every scheduler arg be relative to the training duration is a good restriction. Warmup, especially, likes to run for a fixed absolute amount of steps, and also has weird behavior wrt SSR anyways.